### DISCUSSION TITLE
Do we need to warn users when minor Python versions mismatch?

### DISCUSSION THREAD

**Discussion by orf** [dask/dask#8235] [Ideas]
Dask added minor version mismatch warnings in https://github.com/dask/distributed/issues/595.

I see value in providing this feature, but do we really need to display such a large warning if minor Python versions are mismatched? The Bytecode is explicitly [marked as stable between bugfix releases](https://www.python.org/dev/peps/pep-0006/#prohibitions) and Pickle is also marked with the same stability guarantees.

While rolling out Dask in our company this was a big source of confusion. The error message isn't particularly helpful, and distributions like Debian package Python `3.9.2` rather than the current latest `3.9.7` which makes minor mismatches quite common.

I guess the question is: Are minor mismatches currently such a huge problem that we should frequently display a very large table warning? Maybe a smaller warning on the scheduler only might suffice?

**Comment by jrbourbeau** [comment]
Thanks for raising a discussion item @orf! 

> Are minor mismatches currently such a huge problem that we should frequently display a very large table warning?

Generally speaking software environment mismatches are one of, if not the, most common pain point Dask users run into. However, you bring up a fair point that minor differences in Python are _usually_ okay. That said, there are some commonly used libraries that don't provide the same compatibility guarantees `pickle` does. For example, from [`cloudpickle`s README](https://github.com/cloudpipe/cloudpickle/blob/6099fdb1229adca6b8a013896bc8ec8daabe73b0/README.md): "Cloudpickle can only be used to send objects between the exact same version of Python."

> Maybe a smaller warning on the scheduler only might suffice?

Users don't always check scheduler logs and, given how problematic software version mismatches can be, I think there's value in emitting warnings directly in the user's `Client` Python session. 

> The error message isn't particularly helpful

Do you have any suggestions on how the warning message could be improved? For example, we could add an additional note [here](https://github.com/dask/distributed/blob/8fa3854514c5ede53e9c0381776f6109882ff778/distributed/versions.py#L31-L34) around how often differences in Python versions are problematic.

**Reply by orf** [reply]
> That said, there are some commonly used libraries that don't provide the same compatibility guarantees pickle does. For example, from cloudpickles README: "Cloudpickle can only be used to send objects between the exact same version of Python."

Looking at the history of this message, it appears to come from this comment: https://github.com/cloudpipe/cloudpickle/issues/293#issuecomment-502011167. This is specifically referring to the Bytecode, which is marked as stable _within_ minor versions of Python. So I think that it's referring to the "exact same minor version of python" rather than "the exact same bugfix version of python". 

> Users don't always check scheduler logs and, given how problematic software version mismatches can be, I think there's value in emitting warnings directly in the user's Client Python session.

I totally agree with displaying it in the Clients python session, but we are running a remote dask cluster launched on demand and this is printed on every worker launch. So if you launch 1,000 workers you're going to end up with 5,000 lines of output streamed to you just to display the same table, which isn't great. It's also hard to filter out 😞 

> Do you have any suggestions on how the warning message could be improved? For example, we could add an additional note here around how often differences in Python versions are problematic.

I think there are two ways we could improve this. When you send a cloudpickle object from one version of Python to another, it usually fails to deserialize with a specific `ValueError` mentioning an unknown/invalid opcode. As it's a plain ValueError this isn't simple to explicitly catch, but I think there might be some value in looking at the message and throwing a more informative error rather than one that seems to originate in the guts of the `dis` module.

Secondly, and more generally, we could surface better errors on deserialization if the Python versions don't match. Something like:

```python
try:
    return deserialize()
except Exception as e:
   if not python_versions_match():
       raise DeserializationError("Failed to deserialize message. Python versions don't match, this could be the cause, try with matching versions") from e 
   raise
```

Again this isn't perfect but it puts more context into an exception message that is otherwise quite confusing unless you're quite familiar with Dask.

On the warning message itself, we re-write it to be:

```
Dask detected mismatched versions between the client and workers. Check the table below for significant version differences and be aware these could cause unexpected issues.
```

Which I think is more informative than the current simple `Mismatched versions found`. One of our users told us: "so what? I'm not clear on what this means or what issues it can cause".
